Ensure `EINTR` handling is much more consistent across the board

hamarb123 opened the pull request:
Addresses https://github.com/dotnet/runtime/pull/116639#issuecomment-3043516676

Functions not yet done/checked:
- `select` (I found ~3 candidates for this one, but I'm not certain if they should be changed or not, as they have a timeout parameter)

Functions done are functions with `EINTR` handling in other places, such as `pal_io.c`, and some variants of those functions.

Also addressed lack of partial operation handling for a number of cases.

I haven't checked new usages over the 3 months from the initial commit to where I did the merge.

hamarb123 committed:
Fix handling of `readdir` giving `EINTR`

hamarb123 committed:
Fix handling of `opendir` giving `EINTR`

hamarb123 committed:
Ensure handling of `fcntl` giving `EINTR` is consistent

- Excludes potential changes to libunwind

hamarb123 committed:
Ensure handling of `fsync` giving `EINTR` is consistent

hamarb123 committed:
Ensure handling of `chmod` giving `EINTR` is consistent

hamarb123 committed:
Ensure handling of `ftruncate` giving `EINTR` is consistent

hamarb123 committed:
Ensure handling of `flock` giving `EINTR` is consistent

hamarb123 committed:
Ensure handling of `poll` giving `EINTR` is consistent

I skipped those which were already in infinite loops anyway

hamarb123 committed:
Ensure handling of `waitpid` giving `EINTR` is consistent

- Excludes potential changes to libunwind

hamarb123 committed:
Ensure handling of `getpwuid` giving `EINTR` is consistent

hamarb123 committed:
Ensure handling of `recvmsg` giving `EINTR` is consistent

hamarb123 committed:
Ensure handling of `stat` giving `EINTR` is consistent

- Excludes potential changes to external libs: brotli, libunwind, libtracepoint-control-cpp
- Excludes potential changes to "templates"

hamarb123 committed:
Ensure handling of variants of `stat` giving `EINTR` is consistent

- Excludes potential changes to external lib libunwind
- Excludes potential changes to "templates"

hamarb123 committed:
Ensure handling of `unlink` giving `EINTR` is consistent

- Excludes potential changes to external libs: brotli, libtracepoint-control-cpp

hamarb123 committed:
Ensure handling of `shm_unlink` giving `EINTR` is consistent

hamarb123 committed:
Ensure handling of `pipe{2}` giving `EINTR` is consistent

- Excludes potential changes to external lib libunwind

hamarb123 committed:
Ensure handling of `mkdir` giving `EINTR` is consistent

hamarb123 committed:
Ensure handling of `chdir` giving `EINTR` is consistent

- Excludes potential changes to "templates"

hamarb123 committed:
Ensure handling of `mkfifo` giving `EINTR` is consistent

hamarb123 committed:
Ensure handling of `rename` giving `EINTR` is consistent

hamarb123 committed:
Ensure handling of `rmdir` giving `EINTR` is consistent

hamarb123 committed:
Ensure handling of `ioctl` giving `EINTR` is consistent

- Excludes potential changes to libtracepoint-control-cpp

hamarb123 committed:
Ensure handling of all variants of `statfs` giving `EINTR` is consistent

hamarb123 committed:
Ensure handling of `sendmsg` giving `EINTR` is consistent

hamarb123 committed:
Ensure handling of `accept` giving `EINTR` is consistent

hamarb123 committed:
Ensure handling of `connect` giving `EINTR` is consistent

hamarb123 committed:
Ensure handling of `kevent` giving `EINTR` is consistent

hamarb123 committed:
Ensure handling of `dup2` giving `EINTR` is consistent

hamarb123 committed:
Ensure handling of `remove` giving `EINTR` is consistent

- Skipped tests
- Excludes potential changes to external libs: libeventheader-tracepoint, libunwind

janvorli commented:
@hamarb123 I don't see any notion of possible EINTR error for lseek64, lseek in their doc

hamarb123 commented:
@janvorli I based it off of existing code:
https://github.com/dotnet/runtime/blob/d751cbc4dccf671024760adf5b1c5400cd9eb045/src/native/libs/System.Native/pal_io.c#L765-L782

janvorli commented:
The opendir doc doesn't mention EINTR either.

hamarb123 committed:
Ensure handling of `lseek` & variants giving `EINTR` is consistent

hamarb123 committed:
Ensure handling of `open` & variants giving `EINTR` is consistent

- Excludes potential changes to "templates"

hamarb123 committed:
Ensure we handle `EINTR` & partial operations more consistently for `write` & variants

- Excludes potential changes to some test code
- Excludes potential changes to some external libs

hamarb123 committed:
Fix handling of `closedir` giving `EINTR`

hamarb123 commented:
@jkotas I do plan to finish this PR btw, but have just been a bit busy - if you could un-close it that'd be great, otherwise I can ask someone to re-open it later :)

hamarb123 committed:
Ensure we handle EINTR & partial operations more consistently for pread

hamarb123 committed:
Ensure we handle EINTR & partial operations more consistently for read

- Excludes some potential changes to mono profiler, SystemNative_ReadEvents (seems broken maybe?)

hamarb123 committed:
Merge remote-tracking branch 'upstream/main' into main28

hamarb123 committed:
Fix first set of build failures

hamarb123 committed:
Fix build failures

hamarb123 committed:
Fix build failures

hamarb123 committed:
Fix build failures

hamarb123 committed:
Fix build failures

hamarb123 committed:
Merge remote-tracking branch 'upstream/main' into main28

hamarb123 committed:
Fix build errors

hamarb123 committed:
Merge branch 'main' into main28

hamarb123 committed:
Fix build error

hamarb123 committed:
Fix compiler error

hamarb123 committed:
typo

hamarb123 committed:
Fix build errors

hamarb123 committed:
Remove unrelated whitespace changes VSC added

hamarb123 committed:
Fix compile error

hamarb123 committed:
Remove unrelated whitespace changes VSC added

hamarb123 committed:
Fix compile error

hamarb123 committed:
Merge branch 'main' into main28

hamarb123 committed:
Fix compile error

hamarb123 committed:
Merge branch 'main' into main28
